subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorInjectionIntegrationTest.groovy

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.workers.internal
+
+import org.gradle.api.Project
+import org.gradle.api.internal.file.FileResolver
+import spock.lang.Unroll
+
+class WorkerExecutorInjectionIntegrationTest extends AbstractWorkerExecutorIntegrationTest {
+
+    @Unroll
+    def "workers cannot inject #forbiddenType"() {
+        buildFile << """
+            ${getRunnableInjecting("IsolationMode.NONE", forbiddenType.name)}
+            task runInWorker(type: InjectingWorkerTask)
+        """.stripIndent()
+
+        when:
+        def gradle = executer.withTasks("runInWorker").start()
```

Comment:
We should be able to just use `fails("runInWorker")` here.  I don't think the async execution is necessary.